go-controller/pkg/clustermanager/clustermanager.go

```diff
@@ -0,0 +1,413 @@
+	rl, err := resourcelock.New(
+		resourcelock.LeasesResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		"ovn-kubernetes-cluster-manager",
+		cm.client.CoreV1(),
+		cm.client.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      nodeName,
+			EventRecorder: cm.recorder,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	lec := leaderelection.LeaderElectionConfig{
+		Lock:            rl,
+		LeaseDuration:   time.Duration(config.MasterHA.ElectionLeaseDuration) * time.Second,
+		RenewDeadline:   time.Duration(config.MasterHA.ElectionRenewDeadline) * time.Second,
+		RetryPeriod:     time.Duration(config.MasterHA.ElectionRetryPeriod) * time.Second,
+		ReleaseOnCancel: true,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				klog.Infof("Won leader election; in active mode")
+				// run the cluster controller to init the cluster manager
+				start := time.Now()
+				defer func() {
+					end := time.Since(start)
+					metrics.MetriClusterManagerReadyDuration.Set(end.Seconds())
+				}()
+
+				if err := cm.StartClusterManager(); err != nil {
+					panic(err.Error())
+				}
+				// run the cluster controller to init the master
+				// run only on the active master node.
+				if err := cm.Run(nodeName); err != nil {
+					panic(err.Error())
+				}
+			},
+			OnStoppedLeading: func() {
+				//This node was leader and it lost the election.
+				// Whenever the node transitions from leader to follower,
+				// we need to handle the transition properly like clearing
+				// the cache. It is better to exit for now.
+				// kube will restart and this will become a follower.
+				klog.Infof("No longer leader; exiting")
+				os.Exit(0)
+			},
+			OnNewLeader: func(newLeaderName string) {
+				if newLeaderName != nodeName {
+					klog.Infof("Lost the election to %s; in standby mode", newLeaderName)
+				}
+			},
+		},
+	}
+
+	leaderelection.SetProvider(ovnkubeClusterManagerLeaderMetricsProvider{})
+	leaderElector, err := leaderelection.NewLeaderElector(lec)
+	if err != nil {
+		return err
+	}
+
+	wg.Add(1)
+	go func() {
+		leaderElector.Run(ctx)
+		klog.Infof("Stopped leader election")
+		wg.Done()
+	}()
+
+	return nil
+}
+
+// StartClusterManager runs a subnet IPAM that watches arrival/departure
+// of nodes in the cluster
+// On an addition to the cluster (node create), a new subnet is created for it.
+// ovnkube-master will create the node logical switch and other resources in the
+// OVN Northbound database.
+// Upon deletion of a node, the node subnet is released.
+//
+// TODO: Verify that the cluster was not already called with a different global subnet
+//
+//	If true, then either quit or perform a complete reconfiguration of the cluster (recreate switches/routers with new subnet values)
+func (cm *ClusterManager) StartClusterManager() error {
+	klog.Infof("Starting cluster manager")
+	metrics.RegisterClusterManagerFunctional()
+
+	existingNodes, err := cm.kube.GetNodes()
```

Comment:
`existingNodes` is just used in the following trace, was that for debugging purpose or there is still TODO ?

Reply:
It may have been a copy paste error.

I can remove it if its not required.